File: services/mirror-worker/src/replay/order-stopped.ts
import axios from 'axios';
import { PoolClient } from 'pg';
import { config } from '../config';
import { log } from '../log';
import { ProdUserLog } from '../db/prod';
import { getStagingApiToken } from '../runtime';
import {
  translateEquipmentId,
  translateProductionOrderId,
} from '../translate';

// order-stopped is the explicit stop endpoint (separate from order-changed,
// which is the setup endpoint that also writes a stop semantically). Expected
// payload mirrors StopProductionOrderDto.
//
// Output: POST /api/production-orders/stop
interface OrderStoppedPayload {
  timestamp: string;
  stopType: string;
  idEquipment: number;
  idEnterprise: number;
  idProductionOrder: number;
  productionOrderQuantity: number | string;
}

export async function replayOrderStopped(
  _client: PoolClient,
  row: ProdUserLog,
): Promise<void> {
  const payload = row.payload as OrderStoppedPayload;
  if (!payload || typeof payload !== 'object') {
    throw new Error('order-stopped payload missing or not an object');
  }
  if (typeof payload.idEquipment !== 'number') {
    throw new Error('order-stopped payload missing idEquipment');
  }
  if (typeof payload.idProductionOrder !== 'number') {
    throw new Error('order-stopped payload missing idProductionOrder');
  }

  const stagingEquipmentId = await translateEquipmentId(payload.idEquipment);
  const stagingPoId = await translateProductionOrderId(
    BigInt(payload.idProductionOrder),
  );
  if (stagingPoId === undefined) {
    throw new Error(
      `production_order ${payload.idProductionOrder} unmapped`,
    );
  }

  const body = {
    timestamp: payload.timestamp,
    stopType: payload.stopType,
    idEnterprise: config.stagingEnterpriseId,
    idEquipment: stagingEquipmentId,
    idProductionOrder: Number(stagingPoId),
    productionOrderQuantity: Number(payload.productionOrderQuantity),
  };

  const res = await axios.post(
    `${config.stagingApi.baseUrl}/api/production-orders/stop`,
    body,
    {
      params: {
        token: getStagingApiToken(),
        idEnterprise: config.stagingEnterpriseId,
      },
      headers: {
        'x-user': row.nm_user ?? 'mirror-worker',
        'X-Mirror-Source': config.sourceName,
      },
      validateStatus: () => true,
      timeout: 15_000,
    },
  );

  if (res.status >= 400) {
    throw new Error(
      `staging /api/production-orders/stop returned ${res.status}: ${JSON.stringify(res.data)}`,
    );
  }

  log.info(
    {
      sourceLogId: row.id_user_logs,
      prodPoId: payload.idProductionOrder,
      stagingPoId: stagingPoId.toString(),
      stopType: payload.stopType,
      status: res.status,
    },
    'replayed order-stopped',
  );
}

File: services/mirror-worker-go/internal/replay/order_stopped.go
package replay

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"net/http"

	"github.com/jackc/pgx/v5"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/config"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/db"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/translate"
)

// OrderStopped — same endpoint as order-changed but with the prod payload
// shape that explicitly carries idProductionOrder (rather than the
// oldIdProductionOrder used by order-changed's setup-flow encoding).
type OrderStoppedPayload struct {
	Timestamp               string      `json:"timestamp"`
	StopType                string      `json:"stopType"`
	IDEquipment             int         `json:"idEquipment"`
	IDProductionOrder       int64       `json:"idProductionOrder"`
	ProductionOrderQuantity json.Number `json:"productionOrderQuantity"`
}

func OrderStopped(
	cfg *config.Config,
	t *translate.Translator,
	apiToken func() string,
	httpc *http.Client,
	logger *slog.Logger,
) Handler {
	return func(ctx context.Context, _ pgx.Tx, row db.ProdUserLog) error {
		var p OrderStoppedPayload
		if err := decodeWithNumbers(row.Payload, &p); err != nil {
			return fmt.Errorf("parse order-stopped payload: %w", err)
		}
		if p.IDEquipment == 0 || p.IDProductionOrder == 0 {
			return fmt.Errorf("order-stopped payload missing fields: %+v", p)
		}
		qty, _ := p.ProductionOrderQuantity.Float64()

		stagingEqID, err := t.Equipment(ctx, p.IDEquipment)
		if err != nil {
			return fmt.Errorf("translate equipment: %w", err)
		}
		stagingPOID, err := t.ProductionOrder(ctx, p.IDProductionOrder)
		if err != nil {
			if errors.Is(err, translate.ErrUnmapped) {
				// Same structural-mismatch shape as order-changed / order-started:
				// the PO exists on prod but was never mirrored to staging. No
				// retry will fix this; skip + advance cursor instead of DLQ.
				logger.Info("skipping order-stopped: production_order unmappable on staging",
					slog.Int64("sourceLogID", row.IDUserLogs),
					slog.Int64("prodPOID", p.IDProductionOrder),
					slog.String("translatorErr", err.Error()))
				return fmt.Errorf("production_order %d unmapped (no mirror_id_map row and no id_order business-key match): %w",
					p.IDProductionOrder, ErrSkipReplay)
			}
			return fmt.Errorf("translate production_order: %w", err)
		}

		body := map[string]any{
			"timestamp":               p.Timestamp,
			"stopType":                p.StopType,
			"idEnterprise":            cfg.StagingEnterpriseID,
			"idEquipment":             stagingEqID,
			"idProductionOrder":       stagingPOID,
			"productionOrderQuantity": qty,
		}
		status, _, err := PostStaging(ctx, cfg, httpc, apiToken(), row,
			"/api/production-orders/stop", body)
		if err != nil {
			return err
		}
		logger.Info("replayed order-stopped",
			slog.Int64("sourceLogID", row.IDUserLogs),
			slog.Int64("prodPOID", p.IDProductionOrder),
			slog.Int64("stagingPOID", stagingPOID),
			slog.String("stopType", p.StopType),
			slog.Int("status", status))
		return nil
	}
}
